### ISSUE TITLE
If choosing encryption and Grub is used, prompt on boot is only in US_englisch.

### ISSUE THREAD

**Issue by killajoe** [calamares/calamares#1772]
**If choosing encryption and Grub is used, prompt on boot is only in US_englisch.**
The most common way to setup installs is still using Grub, and if we choose encryption calamares opens input for passwordphrase for LUKS password, on boot up with Grub a lot users have issues because they choose letters and signs not at the same place for non US keyboards and are unable to unlock encryption.

**A warning under the input field about that or a popup with a warning and short description like this:**
![2021-09-04_02-38](https://user-images.githubusercontent.com/16797647/132077161-8fb71724-5adc-452a-bbda-bc506e0c59f3.png)
![2021-09-04_02-46](https://user-images.githubusercontent.com/16797647/132077162-de9e4653-caf7-422e-8da1-39a07b1afd38.png)

**Additional context**
I do not know of any solution to have Grub prompt on full encrypted systems using locale/keyboard settings.
Issue could be on Distributions using systemd-boot or other bootloader implementations so it would be perfect to be able to choose to show that warning or not by setting option, or have implementation to detect if it is needed. (may stupidly complicated)

**Comment by Chrysostomus** [comment]
https://superuser.com/questions/974833/change-the-keyboard-layout-of-grub-in-stage-1

In case someone is interested, this is what it requires to have Grub respect the system keyboard layout for decryption.

I think calamares already supports using the correct layout if your /boot is not encrypted.

**Comment by killajoe** [comment]
> I do not know of any solution to have Grub prompt on full encrypted systems using locale/keyboard settings

yes sure eon full encrypted installs..  what is the most used method. 
So calamares is using grub-mkstandalone ?

**Comment by Chrysostomus** [comment]
No, grub-install command which does not support setting the keyboard layout

**Comment by Chrysostomus** [comment]
in short, this is nor yet a supported feature in calamares

**Comment by apt-ghetto** [comment]
There are already some open issues like  #1726 or #1203 for the very same problem.

Of course, you can workaround the situation with `grub-mkstandalone`, but then you have to live with several downsides:
  * You cannot use Secure Boot (UEFI only) or you have to sign the bootloader yourself (and make sure, that it is really a "Secure Boot" and not only a "Signed Boot")
  * Distros which are executing a `grub-install` after an update may or may not overwrite your standalone grub and/or boot order

**Comment by Chrysostomus** [comment]
These are good points. These issues are not relevant to our distro, but calamares is a crossdirstro tool and has to consider many use cases

**Comment by killajoe** [comment]
My request was also not to implement a solution, and what i wanted to suggest is to simply warn inside installer about the fact that password prompt will be in **us_en**

**Comment by Chrysostomus** [comment]
I think the password prompt in calamares should also set the keyboard to us_en while giving that warning. It's no fun to set a cyrillic password and then discover it is not possible to input it

**Comment by apt-ghetto** [comment]
> I think the password prompt in calamares should also set the keyboard to us_en while giving that warning. It's no fun to set a cyrillic password and then discover it is not possible to input it

Yes, this would be a possible solution, which is probably easy to implement with Calamares. There is only a little problem, when the user tries to unlock the LUKS container with a live medium and a different keyboard layout.

Another approach is to add another LUKS keyslot with the "keyboard layout converted" passphrase. In this case, the user is able to unlock the LUKS container with the en_us and the expected keyboard layout. But I think, this solution needs to be programmed first. Some pseudo code

```c++
auto keyboardLayout = getCurrentKeyboardLayout();
auto passphrase = getPassphrase();
auto convertedPassphrase = keyboardLayout.toEn_Us(passphrase);
addKeyslot(convertedPassphrase);
```

> My request was also not to implement a solution, and what i wanted to suggest is to simply warn inside installer about the fact that password prompt will be in **us_en**

A warning would definitely not hurt.

**Comment by Chrysostomus** [comment]
This would be ideal. The only problem I foresee is that an extra key can increase the time required to decrypt the drive, but I think this acceptable

**Comment by philmmanjaro** [comment]
Well, I think the user is technically fine with having his password converted to en_US for the bootloader. He will still type the same letters he thinks he is typing with the keyboard. Only mess would be when he really uses an American keyboard ...

**Comment by bikepunk** [comment]
I ran into this issue. I installed encrypted system with a passphrase with non us_en characters and now I cannot boot. Have to reinstall and find some workaround.

**Comment by tuxayo** [comment]
> Only mess would be when he really uses an American keyboard ...

Why would this be an issue? Or do you mean switching to any other physical keyboard layout than the one at the installation would cause a mess?

**Comment by logolive** [comment]
I would like to link this fairly compact and likely interoperable solution found on the ArchWiki:

https://wiki.archlinux.org/title/GRUB/Tips_and_tricks#Manual_configuration_of_core_image_for_early_boot

In short, it uses `grub-mkimage` as a middleground to load the keymap, decrypt and mount the LUKS Partition, then delegate to the actual `grub.cfg` generated by many systems using `grub-mkconfig` (or `update-grub`, however it is called)

**Comment by apt-ghetto** [comment]
> In short, it uses `grub-mkimage` as a middleground to load the keymap,

This does not work with Secure Boot.

**Comment by logolive** [comment]
It depends. You can still sign any images and have them be checked before loading (in the end, `grub-install` does something similar to `grub-mkimage` as part of its operation).
But you are then of course unable to ship pre-signed efi binaries, unless you make one for every single keyboard layout.

You are talking above about some distros running `grub-install` after every update. Is there any particular benefit to using this instead of the rather manual approach involving `grub-mkconfig`?

**Comment by apt-ghetto** [comment]
> You can still sign any images and have them be checked before loading

Without the private key it is not possible to sign images.

**Comment by kkofler** [comment]
It is possible to enroll a machine owner key (MOK) in `shim`. But that is (by design) a manual process and not something Calamares can do out of the box.

**Comment by D3vil0p3r** [comment]
I see that the first issue ticket on this was opened 2-3 years ago. @apt-ghetto currently did you apply some solution on your distro for "overcoming" the keymap issue on GRUB when /boot is encrypted?

**Comment by dalto8** [comment]
This is an interesting one.  I would love to implement a solution for this, but I don't see an obvious solution that would apply to Calamares.  Here is the situation as I understand it.

The primary issue is that `grub-install` installs an image without the needed keymaps.

There are three ways to work around that:
1. Create a unified EFI image 
2. Create a custom EFI image that contains the needed keymap
3. Don't use encrypt your kernel/initramfs allowing the decryption to happen later

While it might be interesting for someone to add support to Calamares for 1, it isn't really a solution to the problem as it is really a different way of booting the system altogether.  It also doesn't support secure boot and may not be what most people are expecting.

Creating a custom image as identified in 2 is something that could certainly be implemented in Calamares.  However, it seems fatally flawed to me.  As soon as a user, process or package runs grub-install on their machine the keymap will be missing and they will not be able to decrypt their data with no understanding of why.

Option 3 is already supported by Calamares but it also comes with some cons such as the inability to seamlessly unlock multiple luks partitions.

If there is anything I missed here please point it out.

**Comment by D3vil0p3r** [comment]
Thank you @dalto8 
I don't know if it is related to point 3., but since this problematic situation should occur for cases where `/boot` is encrypted (meaning that the decryption is needed before landing on GRUB), in Calamares, during the installation in the Partition section, could be a good idea to have a "checkbox" for the option "Erase disk" that says "Encrypt also /boot partition"? So, if it is not selected, the `/boot` is not encrypted and we keep GRUB unencrypted so the unlock prompt is moved after GRUB? If it is not clear, please tell me.

**Comment by dalto8** [comment]
> in Calamares, during the installation in the Partition section, could be a good idea to have a "checkbox" for the option "Erase disk" that says "Encrypt also /boot partition"? So, if it is not selected, the `/boot` is not encrypted and we keep GRUB unencrypted so the unlock prompt is moved after GRUB? 

That seems to be too specific to me.  There is no guarantee that the kernel and initramfs are even in `/boot`.  Further, if the distro isn't using grub, encrypting `/boot` might break booting.

**Comment by D3vil0p3r** [comment]
> > in Calamares, during the installation in the Partition section, could be a good idea to have a "checkbox" for the option "Erase disk" that says "Encrypt also /boot partition"? So, if it is not selected, the `/boot` is not encrypted and we keep GRUB unencrypted so the unlock prompt is moved after GRUB?
> 
> That seems to be too specific to me. There is no guarantee that the kernel and initramfs are even in `/boot`. Further, if the distro isn't using grub, encrypting `/boot` might break booting.

Indeed, if Calamares leaves `/boot` unencrypted, it could be the right way, and we should not have the "issue" described in the topic anymore. I'm thinking about this based on https://github.com/calamares/calamares/issues/1025#issuecomment-418279044

**Comment by dalto8** [comment]
> Indeed, if Calamares leaves `/boot` unencrypted, it could be the right way

That isn't a solution for everyone but if it works for your distro it is easy enough to implement by simply changing the ESP location to `/boot`.  Although, it might be interesting to add a feature to allow excluding a partition from encryption.

**Comment by D3vil0p3r** [comment]
> Although, it might be interesting to add a feature to allow excluding a partition from encryption.

Yes, I strongly agree. It could be very useful.

**Comment by logolive** [comment]
Please correct me if I am wrong, but would it not interfere with secure boot if Grub was able to load arbitrary configurations and modules?
I may be missing a config option somewhere, but I do not think grub checks what it loads.
Encryption can provide some integrity guarantees if used correctly,
but the way I understand it, it is being discussed that Grub should not decrypt anything?
This would imply that both the kernel and any initcpios need to stay unencrypted and not integrity protected.
I feel like I am missing something obvious here...

**Comment by dalto8** [comment]
> I feel like I am missing something obvious here...

Well, many of the distros using Calamares don't support secureboot installs so they may have different priorities.  That is part of why there is no "one size fits all" solution to begin with.

**Comment by apt-ghetto** [comment]
> @apt-ghetto currently did you apply some solution on your distro for "overcoming" the keymap issue on GRUB when /boot is encrypted?

No, I did not apply a solution.

**Comment by SylivanKenobi** [comment]
Is there an easy workaround for this? I'm using endeavour with grub and a swiss keyboard layout. It's quiet annoying to enter a complex and save password in a keymap that I'm not used to.

**Comment by dalto8** [comment]
> Is there an easy workaround for this? I'm using endeavour with grub and a swiss keyboard layout.

Some of the workarounds are described above.  None of them are what I would categorize as easy.

The easiest option is to probably to place the boot image at an unencrypted location or use a different bootloader.  There are pros and cons for that though depending on your specific distro.  

It would probably be better to ask about that on the forum for your distro.

**Comment by apt-ghetto** [comment]
> Is there an easy workaround for this?

Easy workarounds are:

- Use only characters that are the same keys on both layouts => But this approach makes brute forcing easier.
- Use another keyslot with the password typed with the us keyboard layout => 3 of 8 keyslots are used for the same container.

The second workaround is my favourite workaround for this problem.

**Comment by SixK** [comment]
Hope to understand the real problem, 
but problem about keyboard in us when trying to unlock partition seem's to be in initramfs.

You have to add a line with KEYMAP=y in /etc/initramfs-tools/initramfs.conf
Then run update-initramfs -u to take modification into account.

Sadly, calamares seem's not to have an option to add key/values to initramfs despite having initramfs and initramfscfg modules.

Hope this tip is sufficient, I also followed some tricks to put grub in another language, but I think my grub config was already doing something for that using locales.

**Comment by dalto8** [comment]
> Hope to understand the real problem,
> but problem about keyboard in us when trying to unlock partition seem's to be in initramfs.

This is about grub's built in decryption when using an encrypted `/boot`.
